Feature/sources type svn support

daniel-brosche opened the pull request:
I have completed a basic implementation of a readonly git svn support based on #151 #164.
In this implementation update,lock and install routines are working with the limitation that git svn repositories currently are always discard changes that are made when using gitman (like an implicit force).
I have tried to make at least as possible changes in the codebase and gathered some ideas about a refactoring to improve extensibility. 

I need some feedback about the current state...

jacebrowning commented on gitman/git.py:
```diff
@@ -15,51 +17,67 @@
+            git('clone', '--reference', reference, repo, os.path.normpath(path))
+    elif type == 'git-svn':
+        # just the preperation for the svn deep clone / checkout here (clone will be made in update function to simplify source.py).
+        os.makedirs(path)
```
I would prefer it if we left the default cause flat rather than indent dozens of lines of code in each affected function:

```
if type == 'git-svn':
    # (new logic)
    return

assert type == `git`
# (current logic)
```

This should also make the diff much smaller.

daniel-brosche replied:
I did adapt the code regarding your advice.

jacebrowning commented on gitman/git.py:
```diff
@@ -15,51 +17,67 @@
+def gitsvn(*args, **kwargs):
+    return call('git', 'svn', *args, **kwargs)
+
+def clean_dir(path):    
```
This could be implemented as another case in the shell utility.

`shell.rm(path, _keep=True)`

daniel-brosche commented:
I'm thinking about how the documentation can be extended regarding git-svn feature.
By default the `type` is `git` so only in the case of importing a svn repository the `type` needs to be configured to `git-svn`. 
Similar to git the git-svn credentials needs to be configured independently beforehand. 
This can be done manually by cloning some svn repo using `git svn clone` and entering the credentials like username and password which then will be stored globally.

Alternatively it is possible to create the credential file by hand, like:

1. Create a file under /home/<username>/.subversion/auth/svn.simple, where the filename is the md5 hash. This is how git svn will find the credentials when challenged.
3. The content of the file will have key value pairs as shown below:

```
K 8
passtype
V 6
simple
K 8
password
V <password character count>
<password>
K 15
svn:realmstring
V 50
<svn url> <SVN Server Name>
K 8
username
V <username character count>
<username>
END
```

jacebrowning commented:
As with git, the actual credential management is outside the scope of this project, but there could be a corresponding document for git-svn that provides recommendations: https://github.com/jacebrowning/gitman/blob/develop/docs/setup/git.md

jacebrowning commented:
@daniel-brosche This is the expected behavior, but I'll update the docs to make this more clear.

daniel-brosche commented:
I will write some notes about credentials using git-svn usage. Is there still something more that I should consider to finish pull request?

jacebrowning commented:
TODO:

- [ ] Fix merge conflicts
- [ ] Add new document about git-svn setup: https://github.com/jacebrowning/gitman/tree/develop/docs/setup
- [ ] Possibly add new document about the git-svn use case: https://github.com/jacebrowning/gitman/tree/develop/docs/use-cases 

This is looking great!

daniel-brosche commented:
Is the documentation ok / understandable?

daniel-brosche commented:
I did fixed a bug in due a merging mistake

jacebrowning commented:
@daniel-brosche The docs seem clear enough to me. You'll also want to add your new documents to `mkdocs.yml` for them to show up.

daniel-brosche commented:
done

jacebrowning commented:
Looks good.

Last thing should be to resolve the linter warnings. You can disable `redefined-builtin` in `.pylint.ini`.

daniel-brosche commented:
Done

daniel-brosche commented:
I have noticed that the gitman tests needs to be adapted in due to the adaption of the git.fetch function (added arguments). I will try to resolve that...

daniel-brosche commented:
I do think that I have fixed a couple but now I'm stuck and need support.

jacebrowning commented:
What can I help you with?

daniel-brosche commented:
I'm not sure how to fix the remaining broken tests.

jacebrowning commented:
It looks like `type` is unset. Maybe change the line to `assert type in ('git', '') for now?

daniel-brosche commented:
Strange, I have added some default assignment which don't seems to work in this case:

```python
    def __init__(self, type, repo, name=None, rev='master', link=None, scripts=None, sparse_paths=None):
        super().__init__()
        self.type = type or 'git'
        self.repo = repo
        self.name = self._infer_name(repo) if name is None else name
        self.rev = rev
        self.link = link
        self.scripts = scripts or []
        self.sparse_paths = sparse_paths or []

        for key in ['name', 'repo', 'rev']:
            if not self[key]:
                msg = "'{}' required for {}".format(key, repr(self))
                raise exceptions.InvalidConfig(msg)
    
        #uncomment next line to print source configuration
        #print(str(self))

    def _on_post_load(self):
        self.type = self.type or 'git'
```

I didn't recognized this problem in my manual tests.
Do you think this is only a problem in the tests?

daniel-brosche commented:
Maybe `_on_post_load` will not be called in the right order.
So when `type` field is not specified in the `gitman.yml` then `_on_post_load` will set the default value `git`.
I have fixed more tests. Could you please fix the last 3 broken tests?
I don't understand how this tests can be affected by my modifications.

daniel-brosche commented:
Are there any crucial problems?

jacebrowning commented:
The `repr()` tests need to be updated for the new format.

daniel-brosche commented:
Can you handle this?

daniel-brosche commented:
Is there some progress on your side or should I try it again. I would really appreciate if we could merge and release this feature, soon...

jacebrowning commented:
I don't have the bandwidth this week. Are you able to reproduce the failing tests locally?

daniel-brosche commented:
Can we continue the merging process? Could you please try to resolve the issues or take a look what the real problem is. I have no clue...

jacebrowning commented:
This is your PR, I cannot finish it for you.

The 3 test failures are reproducible on my machine and `pytest` shows the failure reason for each one.

daniel-brosche commented:
I try to find and fix the problems, again. Also, I need to merge the current state of your develop branch. 

The `make test` fails currently (based on develop branch). 

```sh
Traceback (most recent call last):
  File "/home/osboxes/dev/gitman/.venv/bin/pytest", line 7, in <module>
    from pytest import main
  File "/home/osboxes/dev/gitman/.venv/lib/python3.5/site-packages/pytest.py", line 7, in <module>
    from _pytest.assertion import register_assert_rewrite
  File "/home/osboxes/dev/gitman/.venv/lib/python3.5/site-packages/_pytest/assertion/__init__.py", line 12, in <module>
    from _pytest.assertion import rewrite
  File "/home/osboxes/dev/gitman/.venv/lib/python3.5/site-packages/_pytest/assertion/rewrite.py", line 24, in <module>
    from _pytest.pathlib import fnmatch_ex
  File "/home/osboxes/dev/gitman/.venv/lib/python3.5/site-packages/_pytest/pathlib.py", line 27, in <module>
    from pathlib2 import Path, PurePath
ImportError: No module named 'pathlib2'
Makefile:117: recipe for target 'test-all' failed
make: *** [test-all] Error 1
```

Is Python 3.7 required (I used 3.5)?

jacebrowning commented:
Python 3.5 is supported. Try deleting `.venv` and running the tests again.

daniel-brosche commented:
The problem still exists...

daniel-brosche commented:
I have fixed the tests and also resolved merge conflicts.

At the end the problem regarding `pytest` and the missing `pathlib2` dependency still exists.
I assume that there is an obscure dependency between `python`, `pip` and `pipenv`.
Maybe pathlib2 is part of python 3.7 and therefore it does not need to be addressed by pip on your machine?

As mentioned: on my maschine I have `python 3.5.2` installed and pipenv uses this python version to setup python environment. 
After alot trial and error I found a working `python 3.5.2` configuration.
I have used:
- pipenv 2018.6.25 (`sudo pip install -Iv pipenv==2018.6.25`)
- pip 10.0.1 (`sudo pip install --upgrade pip==10.0.1`)
- added an entry to pipfile (into Test group): `pathlib2 = "*" `

I have tried several other configurations like installing the lastest releases of `pip 18.1` and `pipenv 2018.10.13` with `python 3.5.2` but without success. Besides the problem regarding `pathlib2` I have noticed other problems between `pipenv` and `pip`. It seems that there are some compatiblity issues:

```sh
⟩ make test
pipenv run python setup.py develop
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/pipenv/vendor/requirementslib/_compat.py", line 33, in do_import
    _tmp = importlib.import_module(internal)
  File "/usr/lib/python3.5/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 986, in _gcd_import
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 956, in _find_and_load_unlocked
ImportError: No module named 'pip._internal.basecommand'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/pipenv", line 7, in <module>
    from pipenv import cli
  File "/usr/local/lib/python3.5/dist-packages/pipenv/__init__.py", line 23, in <module>
    from .cli import cli
  File "/usr/local/lib/python3.5/dist-packages/pipenv/cli.py", line 26, in <module>
    from .utils import is_valid_url
  File "/usr/local/lib/python3.5/dist-packages/pipenv/utils.py", line 59, in <module>
    from .vendor.requirementslib import Requirement
  File "/usr/local/lib/python3.5/dist-packages/pipenv/vendor/requirementslib/__init__.py", line 6, in <module>
    from .models import Requirement, Lockfile, Pipfile
  File "/usr/local/lib/python3.5/dist-packages/pipenv/vendor/requirementslib/models/__init__.py", line 8, in <module>
    from .requirements import Requirement
  File "/usr/local/lib/python3.5/dist-packages/pipenv/vendor/requirementslib/models/requirements.py", line 15, in <module>
    from .markers import PipenvMarkers
  File "/usr/local/lib/python3.5/dist-packages/pipenv/vendor/requirementslib/models/markers.py", line 6, in <module>
    from .utils import validate_markers, filter_none
  File "/usr/local/lib/python3.5/dist-packages/pipenv/vendor/requirementslib/models/utils.py", line 9, in <module>
    from .._compat import Link
  File "/usr/local/lib/python3.5/dist-packages/pipenv/vendor/requirementslib/_compat.py", line 55, in <module>
    Command = do_import("basecommand", "Command")
  File "/usr/local/lib/python3.5/dist-packages/pipenv/vendor/requirementslib/_compat.py", line 35, in do_import
    _tmp = importlib.import_module(pip9)
  File "/usr/lib/python3.5/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
ImportError: No module named 'pip.basecommand'
Makefile:53: recipe for target '.venv/.pipenv-c162fb91cd4e944dd8de87df6afab9a2' failed
make: *** [.venv/.pipenv-c162fb91cd4e944dd8de87df6afab9a2] Error 1
```

Maybe we need to define a concrete development environment regarding pip and pipenv besides the current requirements  (version constraints) in the pipfile:
```
[requires]
python_version = "3"
```
I'm not sure if pipenv or poetry is powerful enough to do that? 
For sure a dockerfile can do the job.

Please review my changes...

jacebrowning commented:
I think poetry will be better at managing dependencies for this project.

The change looks good!

The next step is to fix each format/lint warning that you can see when running `$ make check`.

daniel-brosche commented:
I think - that's it..

jacebrowning commented:
Here's the alpha release: https://pypi.org/project/GitMan/1.6a1/
